docs/rfc/rfc-004-accounts.md

```diff
@@ -0,0 +1,323 @@
+  string new_account_kind = 2;
+  google.Protobuf.Any migrate_message = 3;
+}
+
+message MsgMigrateResponse {
+  google.Protobuf.Any data = 1;
+}
+
+```
+
+#### MsgDeploy
+
+Deploys a new instance of the given account `kind` with initial settings represented by the `init_message` which is a `google.Protobuf.Any`.
+Of course the `init_message` can be empty. A response is returned containing the account ID and humanised address, alongside some response
+that the account instantiation might produce.
+
+#### Address derivation
+
+In order to decouple public keys from account addresses, we introduce a new address derivation mechanism which is
+
+
+#### MsgExecute
+
+Sends a `StateTransition` execution request, where the state transition is represented by the `message` which is a `google.Protobuf.Any`.
+The account can then decide if to process it or not based on the `sender`.
+
+### MsgMigrate
+
+Migrates an account to a new version of itself, the new version is represented by the `new_account_kind`. The state transition
+can only be incepted by the account itself, which means that the `sender` must be the account address itself. During the migration
+the account current state is given to the new version of the account, which then executes the migration logic using the `migrate_message`,
+it might change state or not, it's up to the account to decide. The response contains possible data that the account might produce
+after the migration.
+
+#### Authorize Messages
+
+The `Deploy` and `Execute` messages have a field in common called `authorize_messages`, these messages are messages that the account
+can execute on behalf of the sender. For example, in case an account is expecting some funds to be sent from the sender,
+the sender can attach a `MsgSend` that the account can execute on the sender's behalf. These authorizations are short-lived,
+they live only for the duration of the `Deploy` or `Execute` message execution, or until they are consumed.
+
+An alternative would have been to add a `funds` field, like it happens in cosmwasm, which guarantees the called contract that
+the funds are available and sent in the context of the message execution. This would have been a simpler approach, but it would
+have been limited to the context of `MsgSend` only, where the asset is `sdk.Coins`. The proposed generic way, instead, allows
+the account to execute any message on behalf of the sender, which is more flexible, it could include NFT send execution, or
+more complex things like `MsgMultiSend` or `MsgDelegate`, etc.
+
+
+### Further discussion
+
+#### Sub-accounts
+
+We could provide a way to link accounts to other accounts. Maybe during deployment the sender could decide to link the
+newly created to its own account, although there might be use-cases for which the deployer is different from the account
+that needs to be linked, in this case a handshake protocol on linking would need to be defined.
+
+#### Predictable address creation
```

Comment:
We can have predictable addresses be stateless by basing them on `hash(account_kind | account_definition)` where the kind is the registered account implementation as defined in this document, and account_definition a serialization of what that implementation specifies as its unique-account definition.

Some examples:
 * kind=priv/pub_key_accoount, definition=(pubkey: bytes)
 * kind=cosmwasm_account, definition=(contract_addr: string)
 * kind=zk_bearer_account, definition=(proof: string)
 * kind=module_account, definition=(module_name: string)
 * kind=ibc_account, definition=(channel: string, port: string)

Note that in many of these the definition has the same type (a string), so the names on the definition matter. We can use protobuf serialization for this. If a specific account type has collisions, then they can include a nonce in their definition.